risc0/zkp/src/core/blake2b.rs

```diff
@@ -0,0 +1,139 @@
+pub trait Blake2bHasher {
+    /// A function producing a hash from a list of u8.
+    fn blake2b<T: AsRef<[u8]>>(data: T) -> [u8; 32];
+}
+
+/// Implementation of blake2b using CPU.
+pub struct Blake2bImplCpu;
+
+/// Type alias for Blake2b HashSuite using CPU.
+pub type HashSuiteBlake2bCpu = HashSuiteBlake2b<Blake2bImplCpu>;
+
+impl Blake2bHasher for Blake2bImplCpu {
+    fn blake2b<T: AsRef<[u8]>>(data: T) -> [u8; 32] {
+        let mut result = [0; 32];
+        let mut hasher = VarBlake2b::new(32).expect("should create hasher");
+
+        hasher.update(data);
+        hasher.finalize_variable(|slice| {
+            result.copy_from_slice(slice);
+        });
+        result
+    }
+}
+
+/// Blake2b HashSuite.
+/// We are using a generic hasher to allow for different implementations.
+pub struct HashSuiteBlake2b<T: Blake2bHasher> {
+    hasher: PhantomData<T>,
+}
+
+impl<T: Blake2bHasher> HashSuite<BabyBear> for HashSuiteBlake2b<T> {
+    type Hash = ConfigHashBlake2b<T>;
+    type Rng = Blake2bRng<T>;
+}
+
+/// Blake2b ConfigHash.
+pub struct ConfigHashBlake2b<T: Blake2bHasher> {
+    hasher: PhantomData<T>,
+}
+
+impl<T: Blake2bHasher> ConfigHash<BabyBear> for ConfigHashBlake2b<T> {
+    type DigestPtr = Box<Digest>;
+
+    fn hash_pair(a: &Digest, b: &Digest) -> Self::DigestPtr {
+        let concat = [a.as_bytes().as_ref(), b.as_bytes()].concat();
+        Box::new(Digest::from(T::blake2b(concat)))
+    }
+
+    fn hash_elem_slice(slice: &[BabyBearElem]) -> Self::DigestPtr {
+        let mut data = Vec::<u8>::new();
+        for el in slice {
+            data.extend_from_slice(el.as_u32_montgomery().to_be_bytes().as_slice());
+        }
+        Box::new(Digest::from(T::blake2b(data)))
+    }
+
+    fn hash_ext_elem_slice(slice: &[BabyBearExtElem]) -> Self::DigestPtr {
+        let mut data = Vec::<u8>::new();
+        for ext_el in slice {
+            for el in ext_el.subelems() {
+                data.extend_from_slice(el.as_u32_montgomery().to_be_bytes().as_slice());
+            }
+        }
+        Box::new(Digest::from(T::blake2b(data)))
+    }
+}
+
+/// Blake2b-based random number generator.
+pub struct Blake2bRng<T: Blake2bHasher> {
+    current: [u8; 32],
+    hasher: PhantomData<T>,
+}
+
+impl<T: Blake2bHasher> ConfigRng<BabyBear> for Blake2bRng<T> {
+    fn new() -> Self {
+        Self {
+            current: [0; 32],
+            hasher: Default::default(),
+        }
+    }
+
+    fn mix(&mut self, val: &Digest) {
+        let concat = [self.current.as_ref(), val.as_bytes()].concat();
+        self.current = T::blake2b(concat);
+    }
+
+    fn random_u32(&mut self) -> u32 {
+        let next = T::blake2b(self.current);
+        self.current = next;
+
+        ((next[0] as u32) << 24) +
+            ((next[1] as u32) << 16) +
+            ((next[2] as u32) << 8) +
+            ((next[3] as u32) << 0)
+    }
+
+    fn random_elem(&mut self) -> BabyBearElem {
+        BabyBearElem::new(self.random_u32())
```

Comment:
When the BB element is constructed here, the uniformly distribute value x \in [0, 2^32) is moved into the range [0, P) via x % P, which will result in a significantly non-uniform distribution. as some elements appear 3 times in [0, 2^32) % P, and some appear only 2 times.  This non-uniformity breaks the assumptions used in the soundness arguments for the proof system.  You can either use a technique similar to the random_elem implementation in Sha2 which produces something cryptographically close to uniform, or you can use rejection sampling (https://en.wikipedia.org/wiki/Rejection_sampling).  Rejection sampling is probably  easier and more efficient on average in terms of samples drawn, the reason we use an alternate technique in the Sha2 implementation is to make sure the number of samples is always fixed, which matters for certain recursion use cases.

Reply:
Thanks for pointing me in the direction I should go, it's really helpful!

Reply:
Yup, looks good.